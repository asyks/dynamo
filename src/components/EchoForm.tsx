import React from 'react'

import '../styles/App.css'

export interface Props {
  connect: () => void
  disconnect: () => void
  send: () => void
}

export const EchoForm: React.FC<Props> = props => {
  let [messages, setMessages] = React.useState([])

  const logElems: JSX.Element[] = []
  messages.forEach((value, i) => {
    logElems.push(
      <li key={i}>{value}</li>
    )
  })

  return (
    <div>
      <div>websocket echo</div>
      <button onClick={
        () => {
          const message = props.connect()
          setMessages(messages.concat([message] as never[]))
        }
      }>Connect</button>
      <button onClick={
        () => {
          const message = props.disconnect()
          setMessages(messages.concat([message] as never[]))
        }
      }>Disconnect</button>
      <input type="text" id="message" />
      <button onClick={
        () => {
          const message = props.send()
          setMessages(messages.concat([message] as never[]))
        }
      }>Send</button>
      <ul className="Web-socket-echo-log" id="log">{logElems}</ul>
    </div>
  )
}

