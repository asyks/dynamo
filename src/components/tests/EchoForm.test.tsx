import React from 'react'
import renderer from 'react-test-renderer'
import { fireEvent, render } from '@testing-library/react'

import EchoForm, { defaultMessage } from '../EchoForm'

describe("components/EchoForm", () => {
  test("renders with noop handlers", () => {
    const noop = jest.fn()

    expect(renderer.create(
      <EchoForm connect={noop} disconnect={noop} send={noop} />
    ).toJSON()).toMatchSnapshot()
  })

  test("registers button clicks, and text input", () => {
    const mockConnect = jest.fn()
    const mockDisconnect = jest.fn()
    const mockSend = jest.fn()
    const { container, getByText, getByDisplayValue } = render(
      <EchoForm
        connect={mockConnect}
        disconnect={mockDisconnect}
        send={mockSend}
      />
    )
    fireEvent.click(getByText("Connect"))
    fireEvent.change(
      getByDisplayValue(defaultMessage), { target: { value: "Test Message" } }
    )
    fireEvent.click(getByText("Send"))
    fireEvent.click(getByText("Disconnect"))

    expect(mockConnect.mock.calls.length).toBe(1)
    expect(mockDisconnect.mock.calls.length).toBe(1)
    expect(mockSend.mock.calls.length).toBe(1)
  })
})
