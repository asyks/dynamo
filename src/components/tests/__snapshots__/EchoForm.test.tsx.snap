// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/EchoForm renders with noop handlers 1`] = `
<div>
  <div>
    websocket echo
  </div>
  <button
    onClick={[Function]}
  >
    Connect
  </button>
  <input
    onChange={[Function]}
    type="text"
    value="HelloWebSocket!"
  />
  <button
    onClick={[Function]}
  >
    Send
  </button>
  <button
    onClick={[MockFunction]}
  >
    Disconnect
  </button>
  <ul
    className="Web-socket-echo-log"
  />
</div>
`;
